cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -0,0 +1,739 @@
+    field("date64", date64()),
+    field("str", utf8()),
+    field("list_i32", list(int32())),
+    field("struct", struct_({
+                        field("i32", int32()),
+                        field("str", utf8()),
+                        field("struct_i32_str",
+                              struct_({field("i32", int32()), field("str", utf8())})),
+                    })),
+    field("list_struct", list(struct_({
+                             field("i32", int32()),
+                             field("str", utf8()),
+                             field("struct_i32_str", struct_({field("i32", int32()),
+                                                              field("str", utf8())})),
+                         }))),
+    field("dict_str", dictionary(int32(), utf8())),
+    field("dict_i32", dictionary(int32(), int32())),
+    field("ts_ns", timestamp(TimeUnit::NANO)),
+});
+
+std::shared_ptr<DataType> StripFieldNames(std::shared_ptr<DataType> type) {
+  if (type->id() == Type::STRUCT) {
+    FieldVector fields(type->num_fields());
+    for (int i = 0; i < type->num_fields(); ++i) {
+      fields[i] = type->field(i)->WithName("");
+    }
+    return struct_(std::move(fields));
+  }
+
+  if (type->id() == Type::LIST) {
+    return list(type->field(0)->WithName(""));
+  }
+
+  return type;
+}
+
+// map to an index-only field reference
+inline FieldRef BoringRef(FieldRef ref) {
+  auto path = *ref.FindOne(*kBoringSchema);
+  return {std::move(path)};
+}
+
+inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
+  if (expr.literal()) return expr;
+
+  if (auto ref = expr.field_ref()) {
+    return compute::field_ref(*ref->FindOne(*kBoringSchema));
+  }
+
+  auto modified_call = *CallNotNull(expr);
+  for (auto& arg : modified_call.arguments) {
+    arg = UseBoringRefs(arg);
+  }
+  return compute::Expression{std::move(modified_call)};
+}
+
+std::shared_ptr<Buffer> SubstraitFromJSON(util::string_view type_name,
+                                          util::string_view json) {
+  auto maybe_buf = internal::SubstraitFromJSON(type_name, json);
+  DCHECK_OK(maybe_buf.status());
+  return maybe_buf.ValueOrDie();
+}
+
+std::string SubstraitToJSON(util::string_view type_name, const Buffer& buf) {
+  auto maybe_buf = internal::SubstraitToJSON(type_name, buf);
+  DCHECK_OK(maybe_buf.status());
+  return maybe_buf.ValueOrDie();
+}
+
+TEST(Substrait, SupportedTypes) {
+  auto ExpectEq = [](util::string_view json, std::shared_ptr<DataType> expected_type) {
+    ARROW_SCOPED_TRACE(json);
+
+    ExtensionSet empty;
+    auto buf = SubstraitFromJSON("Type", json);
+    ASSERT_OK_AND_ASSIGN(auto type, DeserializeType(*buf, empty));
+
+    EXPECT_EQ(*type, *expected_type);
+
+    ASSERT_OK_AND_ASSIGN(auto serialized, SerializeType(*type, &empty));
+    EXPECT_EQ(empty.types().size(), 0);
+
+    // FIXME chokes on NULLABILITY_UNSPECIFIED
+    // EXPECT_THAT(internal::CheckMessagesEquivalent("Type", *buf, *serialized), Ok());
```

Comment:
Can you remove this, fix it, or add a JIRA for the follow up work?

Reply:
I started working on doing this, but the diff became such that it's probably better to split it up at this stage. The crux is that `NULLABILITY_UNSPECIFIED` is (I think) illegal, but currently abused by `SerializeType` to get around the fact that Arrow types do not have the intrinsic nullability tag that Substrait types do. See ARROW-15618

Reply:
Thanks for creating the JIRA.